From 6dbfbcc45b81790a0e198cc9f76c82fa3411c2aa Mon Sep 17 00:00:00 2001
From: llin4 <ling.lin@intel.com>
Date: Thu, 29 Nov 2012 14:27:09 +0800
Subject: [PORT FROM MAIN] libcore: houdini hook rebase on MR1

BZ: 71657

libcore houdini hook rebase on MR1

Change-Id: I983db5c2584ccdd1f0ba6e1664e4e255b8c374ef
Orig-Change-Id: I4dbd2bef4d0da5c943aa90b4f7c91bdf5b5cb860
Signed-off-by: llin4 <ling.lin@intel.com>
---
 NativeCode.mk                             |    5 +++++
 luni/src/main/native/libcore_io_Posix.cpp |    8 ++++++++
 2 files changed, 13 insertions(+), 0 deletions(-)

diff --git a/NativeCode.mk b/NativeCode.mk
index b7c8b0e..42f5d2d 100644
--- a/NativeCode.mk
+++ b/NativeCode.mk
@@ -118,6 +118,11 @@ LOCAL_ADDITIONAL_DEPENDENCIES := $(LOCAL_PATH)/NativeCode.mk
 LOCAL_C_INCLUDES += external/stlport/stlport bionic/ bionic/libstdc++/include
 LOCAL_SHARED_LIBRARIES += libstlport
 
+ifeq ($(INTEL_HOUDINI), true)
+    LOCAL_CPPFLAGS += -DWITH_HOUDINI
+    LOCAL_STATIC_LIBRARIES += houdini_hook
+endif
+
 include $(BUILD_SHARED_LIBRARY)
 
 
diff --git a/luni/src/main/native/libcore_io_Posix.cpp b/luni/src/main/native/libcore_io_Posix.cpp
index 1daa83d..781d45a 100644
--- a/luni/src/main/native/libcore_io_Posix.cpp
+++ b/luni/src/main/native/libcore_io_Posix.cpp
@@ -937,12 +937,20 @@ static void Posix_munmap(JNIEnv* env, jobject, jlong address, jlong byteCount) {
     throwIfMinusOne(env, "munmap", TEMP_FAILURE_RETRY(munmap(ptr, byteCount)));
 }
 
+#ifdef WITH_HOUDINI
+extern "C" int houdini_hook_open(const char *path, int flags, int mode);
+#endif
+
 static jobject Posix_open(JNIEnv* env, jobject, jstring javaPath, jint flags, jint mode) {
     ScopedUtfChars path(env, javaPath);
     if (path.c_str() == NULL) {
         return NULL;
     }
+#ifdef WITH_HOUDINI
+    int fd = throwIfMinusOne(env, "open", houdini_hook_open(path.c_str(), flags, mode));
+#else
     int fd = throwIfMinusOne(env, "open", TEMP_FAILURE_RETRY(open(path.c_str(), flags, mode)));
+#endif
     return fd != -1 ? jniCreateFileDescriptor(env, fd) : NULL;
 }
 
-- 
1.7.4.1

